### DISCUSSION TITLE
Little suggestions

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#133] [General]
Currently running nested loops allows for getting the current index of the outer most loop in the innermost loop for two levels.
In life many things are tripled: 3D, RGB, Three rings for the elven kin...sorry.
If it is not too much effort or has some bad impact I didn't recognize, I would like to suggest i,j and k as indices
availble in the innermost level of nested loops.

The i2c scanner ( in zeptoforth/test/rp2040/. >> great software! :) "saves my life" just a few minutes ago) prints out a lot of "I2C target address not acknowledged" messages. Really interesting are the addresses, which were found, instead - I think. May be printing only those makes the scanner faster and the given information more rich. 

Question: I want to talk to  a DS3231 I2C device on $68. What parts of the raw protocol as seen in the datasheet of the DS3231 are needed to be include into the buffer to be sent from master (pico) to the slave (DS3231) and which are inserted by Zeptoforth or the i2c hardware of the pico automagically?

**Comment by tabemann** [comment]
On Wed, Aug 14, 2024 at 12:38 PM mcctuxic ***@***.***> wrote:

> Currently running nested loops allows for getting the current index of the
> outer most loop in the innermost loop for two levels.
> In life many things are tripled: 3D, RGB, Three rings for the elven
> kin...sorry.
> If it is not too much effort or has some bad impact I didn't recognize, I
> would like to suggest i,j and k as indices
> availble in the innermost level of nested loops.
>
You could use:

0 { k }
begin k limit u< while
  ...
  1 +to k
repeat

or like. Note that in zeptoforth you can use locals defined outside DO
LOOP's inside them,

> The i2c scanner ( in zeptoforth/test/rp2040/. >> great software! :) "saves
> my life" just a few minutes ago) prints out a lot of "I2C target address
> not acknowledged" messages. Really interesting are the addresses, which
> were found, instead - I think. May be printing only those makes the scanner
> faster and the given information more rich.
>
I could make it print only the acknowledged addresses, but it wouldn't make
it any faster.

> Question: I want to talk to a DS3231 I2C device on $68. What parts of the
> raw protocol as seen in the datasheet of the DS3231 are needed to be
> include into the buffer to be sent from master (pico) to the slave (DS3231)
> and which are inserted by Zeptoforth or the i2c hardware of the pico
> automagically?
>
 As I haven't read the DS3231 datasheet offhand I wouldn't know, but a good
source of this kind of info would be to look at the refman of the RP2040
(or if you plan on targeting the RP2350, its refman).

Travis
